### PR TITLE
add gamepad support through SDL2

### PR DESCRIPTION
**Pull request by pigoz**
If someone could test on Linux and Windows that would be great! I tested it on OSX with a Dualshock 4

The code is very basic:

- only handles gamepads, could be extended for generic joysticks in the
  future.
- only has button mappings for controllers natively supported by SDL2.
  I heard more can be added through env vars, there's also ways to load
  mappings from text files, but I'd rather not go there yet. Common ones
  like Dualshock are supported natively.
- analog buttons like L2/R2 and AXIS are mapped to discrete buttons with
  a treshold.

I tested on OSX, should work elsewhere!

Ideally, you shouldn't need enter any text here, and your commit messages should
explain your changes sufficiently (especially why they are needed). Read
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md for coding
style and development conventions. Remove this text block, but if you haven't
agreed to it before, leave the following sentence in place:

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

**Comment by NiLuJe** [comment]
I'm getting the appropriate key events on Linux w/ a DS4v2, will try actually binding them to something useful later ;).

File: input/gamepad.c [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        MP_KEY_GAMEPAD_L2 | MP_KEY_STATE_DOWN },
+
+    { SDL_CONTROLLER_AXIS_TRIGGERRIGHT,
+        INVALID_KEY,
+        MP_KEY_GAMEPAD_R2 | MP_KEY_STATE_UP,
+        MP_KEY_GAMEPAD_R2 | MP_KEY_STATE_DOWN },
+};
+
+static int lookup_button_mp_key(int sdl_key) {
+    for (int i = 0; MP_ARRAY_SIZE(button_map); i++) {
```
Comment (sfan5) [inline input/gamepad.c]: `i < MP_ARRAY_SIZE(button_map)`

File: input/gamepad.c [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    const int positive = 3;
+
+    const float threshold = (sdl_axis_max * 0.01);
+    const int state =
+        value >= sdl_axis_max - threshold ?
+        positive :
+            value <= threshold - sdl_axis_max ?
+            negative : neutral;
+
+    for (int i = 0; MP_ARRAY_SIZE(axis_map); i++) {
```
Comment (sfan5) [inline input/gamepad.c]: `i < MP_ARRAY_SIZE(button_map)`

File: input/gamepad.c [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    const int negative = 1;
+    const int neutral = 2;
+    const int positive = 3;
+
+    const float threshold = (sdl_axis_max * 0.01);
+    const int state =
+        value >= sdl_axis_max - threshold ?
+        positive :
+            value <= threshold - sdl_axis_max ?
+            negative : neutral;
```
Comment (sfan5) [inline input/gamepad.c]: this is kinda hard to read, an if-else cascade would be better IMO

**Comment by pigoz** [comment]
@sfan5 thanks 👍  fixed everything

**Comment by rld243** [comment]
Can confirm it works just fine on Windows 10 with Logitech F310. Tried both XInput and DirectInput modes.

**Comment by pigoz** [comment]
@rld243 thank you very much for your time!

> Maybe rename cross, square, circle, triangle buttons to something generic like button1, button2 ..? XInput gamepads have A X B Y buttons instead of them. Other buttons are also named differently (LB LT RB RT).

SDL also uses what to happens to be the Xbox/XInput naming convention. I'd be willing to change L1/R1/L2/R2/L3/R3 to the more neutral Xinput conuterparts, i.e.:

```
L1 = LEFT_SHOULDER
L2 = LEFT_TRIGGER
L3 = LEFT_STICK
```

On the other hand A X B Y, is super jarring with what is the most widespread controller (Dualshock). Maybe it could be changed to something that works with both Sony and Microsoft conventions, such as the direction of the buttons. Since the dpad ones have a `DPAD_` prefix, I'd go with:

````
Triangle: ACTION_UP
Square: ACTION_LEFT
Cross: ACTION_DOWN
Circle: ACTION_RIGHT
```

or something similar (other prefixes could be better, I'd like suggestions!).

**Comment by pigoz** [comment]
@rld243 right, I got them mixed up because I was looking at my Nintendo 3ds where X/Y and B/A are reversed when compared to Xbox controllers.

Another reason not to use the ABXY naming and choose something more general 👍

**Comment by rld243** [comment]
@pigoz 
Sure, there are even some gamepads which have numbers on them instead of letters or symbols.

**Comment by pigoz** [comment]
Changed the names as outlined in my previous message

**Comment by sfan5** [comment]
Got around to testing this, it works perfectly expect that repeating keypresses don't seem to work reliably.

With `GAMEPAD_LEFT_STICK_RIGHT` bound to seek, holding the stick down to the right will not continue to seek (though sometimes it works).
Same thing with `GAMEPAD_LEFT_TRIGGER`, except it usually works but sometimes doesn't.
This is with an USB-attached Xbox360 controller, in case it matters.

**Comment by pigoz** [comment]
Yeah, repeats are broken with axis buttons (SDL kinda generates random-ish events when holding a direction)

**Comment by wm4** [comment]
Status?

**Comment by pigoz** [comment]
I still haven't figured out why Analog AXIS are broken. To be honest, I didn't even check because I'm using only the discrete buttons. It might be SDL's fault.

**Comment by pigoz** [comment]
Everything should be solved now. If someone can maybe test it on Windows with Xbox controller that would be cool. Axis and Trigger key repeat should work correctly (it does here OSX + Dualshock 4).

I used the following cli `./build/mpv --no-config --input-conf=test/input-gamepad.conf --msg-level=input=trace`

File: input/sdl_gamepad.c [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    atomic_store(&p->cancel_requested, false);
+    src->priv = p;
+    src->cancel = request_cancel;
+    src->uninit = uninit;
+    mp_input_src_init_done(src);
+
+    SDL_Event ev;
+
+    while (!atomic_load(&p->cancel_requested)) {
+        while (SDL_PollEvent(&ev) != 0) {
```
Comment (wm4) [inline input/sdl_gamepad.c]: It looks like this will just burn a core continuously.

File: input/sdl_gamepad.c [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    // must be called on the same thread of SDL_InitSubSystem, so uninit
+    // callback can't be used for this
+    SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+}
+
+void mp_input_sdl_gamepad_add(struct input_ctx *ictx)
+{
+    mp_input_add_thread_src(ictx, NULL, read_gamepad_thread);
```
Comment (wm4) [inline input/sdl_gamepad.c]: I planned to kill this API, only the win32 pipe stuff still used it. I may request that you adjust this code some time in the future when it comes to it.

File: input/sdl_gamepad.c (hunk 4 above)
Comment (pigoz) [reply input/sdl_gamepad.c]: Thanks, fixed. That's why my room was awfully hot when using mpv in the past months.

File: input/sdl_gamepad.c (hunk 5 above)
Comment (pigoz) [reply input/sdl_gamepad.c]: Ok, I actually use this feature.

File: input/sdl_gamepad.c [hunk 6]
```diff
@@ -160,6 +161,13 @@ static void read_gamepad_thread(struct mp_input_src *src, void *param)
 
     SDL_GameController *controller;
     struct priv *p = talloc_zero(src, struct priv);
+    p->cancel_wakeup = SDL_RegisterEvents(1);
```
Comment (wm4) [inline input/sdl_gamepad.c]: Unfortunately, SDL_RegisterEvents is a global thing. Events can't be unregistered, so every initialization of the gamepad code will consume one more global event. I think you don't need to care about this. If you do, use pthread_once to register a single global event.

**Comment by rld243** [comment]
@pigoz 
Just tested again with Logitech F310 on Windows 10. AXIS/Trigger keys repeats and everything else seem to be working as intended.

File: input/sdl_gamepad.c [hunk 7]
```diff
@@ -17,15 +17,20 @@
...
 #include "common/common.h"
 #include "osdep/atomic.h"
 #include "common/msg.h"
 #include "input.h"
 #include "input/keycodes.h"
 
-struct priv {
-    Uint32 cancel_wakeup;
-};
+Uint32 gamepad_cancel_wakeup;
```
Comment (wm4) [inline input/sdl_gamepad.c]: Forgot static?

File: input/sdl_gamepad.c [hunk 8]
```diff
@@ -17,15 +17,20 @@
...
+
+static void initalize_events() {
```
Comment (wm4) [inline input/sdl_gamepad.c]: Coding style.

File: input/sdl_gamepad.c (hunk 7 above)
Comment (pigoz) [reply input/sdl_gamepad.c]: yes, I can't see warnings from all the clang spam

**Comment by pigoz** [comment]
I added support for device hotplugging, in an ugly way. Let me know if you find it acceptable.

File: input/sdl_gamepad.c [hunk 9]
```diff
@@ -157,6 +162,49 @@ static void uninit(struct mp_input_src *src)
...
+        }
+    }
+
+    return false;
+}
+
+static void remove_gamepad(struct mp_input_src *src, int id)
+{
+    if (id >= MAX_CONTROLLERS) {
+        MP_WARN(src, "can't remove controller id %d, out bounds", id);
```
Comment (sfan5) [inline input/sdl_gamepad.c]: missing return

Comment (pigoz) [reply input/sdl_gamepad.c]: woops. thanks

**Comment by graphitemaster** [comment]
I have a couple concerns with this PR:

I think the approach taken with a separate input thread here isn't safe. SDL2 only lightly wraps the OS event loop and on platforms like Windows and macOS, this means event polling has to be done on the main thread. The SDL2 documentation even mentions this in it's official documentation here: <https://wiki.libsdl.org/CategoryThread>

>  NOTE: You should not expect to be able to create a window, render, or receive events on any thread other than the main one.

The management of gamepads seems unsafe too, a fixed size of 32 elements may be appropriate in that no one would ever plug in more than 32 controllers, until you consider that the id space is monotonically increasing even for hot swapping and reconnections, e.g: `<SDL_joystick.h>`
```c
/**
 * This is a unique ID for a joystick for the time it is connected to the system,
 * and is never reused for the lifetime of the application. If the joystick is
 * disconnected and reconnected, it will get a new ID.
 *
 * The ID value starts at 0 and increments from there. The value -1 is an invalid ID.
 */
typedef Sint32 SDL_JoystickID;
```
This is quite easy to overflow when the controller is low on battery and is constantly reconnecting (in the case of wireless gamepads).

The handling of trigger buttons seem incorrect in `lookup_analog_mp_key` which does `state > 0 ? positive_neutral : negative_neutral` to determine which state to be in, unfortunately, the handling of shoulder buttons (triggers) have a different range than the axis, as documented here: <https://wiki.libsdl.org/SDL_GameControllerGetAxis>, in particular:

> The state is a value ranging from -32768 to 32767. Triggers, however, range from 0 to 32767 (they never return a negative value). 

So every trigger button is treated as `negative_neutral` if I'm reading this correctly.

I'm not sure if I like that these constants are baked in and non-configurable:
```c
const float activation_threshold = sdl_axis_max * 0.33;
const float noise_threshold = sdl_axis_max * 0.06;
```
These values are highly dependent on the quality of your controller and calibration within the operating system. Can these be exposed to the general mpv property framework maybe with sane defaults (like the above) set as defaults?

Note: In this PR, it appears the handling of open gamepads is only necessary for the purposes of knowing which to open for events and close, the events themselves are not be routed on a "per-controller" basis, so if you had multiple controllers plugged in, they all do the same and you can't have binds in mpv that only apply to one controller for instance (not sure if this was intended, just something to note for others reading.)

**Comment by pigoz** [comment]
> I think the approach taken with a separate input thread here isn't safe. SDL2 only lightly wraps the OS event loop and on platforms like Windows and macOS, this means event polling has to be done on the main thread.

I don't know about Windows, but on macOS you can definitely pump events on non-main threads (if they are not windowing related). mpv has it's own loop on the main thread, and rearchitecting mpv to better accomodate SDL would be quite hard. FWIW vo_sdl crashes on OSX because it creates windows on the vo thread.

> This is quite easy to overflow when the controller is low on battery and is constantly reconnecting (in the case of wireless gamepads).

I should probably allow a single controller connected, there's no way in mpv's event system to send different events from different controllers (event names in mpv are static).

> These values are highly dependent on the quality of your controller and calibration within the operating system. Can these be exposed to the general mpv property framework maybe with sane defaults (like the above) set as defaults?

I can expose them, if others are not against.

> so if you had multiple controllers plugged in, they all do the same and you can't have binds in mpv that only apply to one controller for instance (not sure if this was intended, just something to note for others reading.)

Yes, it's intended, in mpv we can just controller buttons to "mpv buttons". It's not clear to me how we would go about configuring mappings for single controllers. To make it more practical, we could force a single controller active at all times, and users can change mapping by changing their input.conf

**Comment by pigoz** [comment]
Changed my code to only allow one concurrently connected controller. I don't think it makes sense to introduce the complexity of handling more than one controller (mappings would be shared anyway).

As for the event loop... nothing I can do about it. For all I care this is ready to be merged now.

Also rebased against master.

**Comment by wm4** [comment]
Push it, I guess. The commit message should probably have "input: " as prefix.

**Comment by pigoz** [comment]
Merged. Thank you for the reviews!

**Pull request merged**